test/utils.go

```diff
@@ -0,0 +1,338 @@
+	for i := 0; i < num; i++ {
+		cis[i] = &agent.ContainerInstanceArn
+	}
+
+	resp, err := ECS.StartTask(&ecs.StartTaskInput{
+		Cluster:            &agent.Cluster,
+		ContainerInstances: cis,
+		TaskDefinition:     &td,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Failures) != 0 || len(resp.Tasks) == 0 {
+		return nil, errors.New("Failure starting task: " + *resp.Failures[0].Reason)
+	}
+
+	testTasks := make([]*TestTask, num)
+	for i, task := range resp.Tasks {
+		testTasks[i] = &TestTask{task}
+	}
+	return testTasks, nil
+}
+
+func (agent *TestAgent) StartTask(t *testing.T, task string) (*TestTask, error) {
+	tasks, err := agent.StartMultipleTasks(t, task, 1)
+	if err != nil {
+		return nil, err
+	}
+	return tasks[0], nil
+}
+
+type TestTask struct {
+	*ecs.Task
+}
+
+func (task *TestTask) waitStatus(timeout time.Duration, status string) error {
+	timer := time.NewTimer(timeout)
+	atStatus := make(chan error)
+
+	cancelled := false
+	go func() {
+		if *task.LastStatus == "STOPPED" && status != "STOPPED" {
+			atStatus <- errors.New("Task terminal; will never reach " + status)
+			return
+		}
+		for *task.LastStatus != status && !cancelled {
+			res, err := ECS.DescribeTasks(&ecs.DescribeTasksInput{
+				Cluster: task.ClusterARN,
+				Tasks:   []*string{task.TaskARN},
+			})
+			if err == nil && len(res.Failures) == 0 {
+				task.Task = res.Tasks[0]
+			}
+			if *task.LastStatus == status {
+				break
+				return
+			}
+			if *task.LastStatus == "STOPPED" && status != "STOPPED" {
+				atStatus <- errors.New("Task terminal; will never reach " + status)
+			}
+			time.Sleep(5 * time.Second)
+		}
+		atStatus <- nil
+	}()
+
+	select {
+	case err := <-atStatus:
+		return err
+	case <-timer.C:
+		cancelled = true
+		return errors.New("Timed out waiting for task to reach" + status + ": " + *task.TaskDefinitionARN + ", " + *task.TaskARN)
+	}
+}
+
+func (task *TestTask) ContainerExitcode(name string) (int, bool) {
+	for _, cont := range task.Containers {
+		if cont != nil && cont.Name != nil && cont.ExitCode != nil {
+			if *cont.Name == name {
+				return int(*cont.ExitCode), true
+			}
+		}
+	}
+	return 0, false
+}
+
+func (task *TestTask) WaitRunning(timeout time.Duration) error {
+	return task.waitStatus(timeout, "RUNNING")
+}
+
+func (task *TestTask) WaitStopped(timeout time.Duration) error {
+	return task.waitStatus(timeout, "STOPPED")
+}
+
+func (task *TestTask) ExpectErrorType(containerName, errType string, timeout time.Duration) error {
+	task.WaitStopped(timeout)
+
+	for _, container := range task.Containers {
+		if *container.Name != containerName {
+			continue
+		}
+		if container.Reason == nil {
+			return errors.New("Expected error reason")
+		}
+		errParts := strings.SplitN(*container.Reason, ":", 2)
+		if len(errParts) != 2 {
+			return errors.New("Error did not have a type: " + *container.Reason)
+		}
+		if errParts[0] != errType {
+			return errors.New("Type did not match: " + *container.Reason)
```

Comment:
Could we add both the expectation as well as the actual in the error string? "i.e. Type did not match. Expected x, got y. Reason".